contracts/atomic-swap/src/contract.rs

```diff
@@ -0,0 +1,619 @@
+use cosmwasm_std::{
+    log, to_binary, Api, BankMsg, Binary, Coin, CosmosMsg, Env, Extern, HandleResponse, HumanAddr,
+    InitResponse, Querier, StdError, StdResult, Storage,
+};
+use sha2::{Digest, Sha256};
+
+use cw2::{set_contract_version, ContractVersion};
+
+use crate::msg::{
+    is_valid_name, CreateMsg, DetailsResponse, HandleMsg, InitMsg, ListResponse, QueryMsg,
+};
+use crate::state::{all_swap_ids, atomic_swaps, atomic_swaps_read, AtomicSwap};
+
+// Version info, for migration info
+const CONTRACT_NAME: &str = "crates.io:atomic-swap";
+const CONTRACT_VERSION: &str = "v0.1.0";
+
+pub fn init<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    _env: Env,
+    _msg: InitMsg,
+) -> StdResult<InitResponse> {
+    let version = ContractVersion {
+        contract: CONTRACT_NAME.to_string(),
+        version: CONTRACT_VERSION.to_string(),
+    };
+    set_contract_version(&mut deps.storage, &version)?;
+    // No setup
+    Ok(InitResponse::default())
+}
+
+pub fn handle<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    msg: HandleMsg,
+) -> StdResult<HandleResponse> {
+    match msg {
+        HandleMsg::Create(msg) => try_create(deps, env, msg),
+        HandleMsg::Release { id, preimage } => try_release(deps, env, id, preimage),
+        HandleMsg::Refund { id } => try_refund(deps, env, id),
+    }
+}
+
+pub fn try_create<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    msg: CreateMsg,
+) -> StdResult<HandleResponse> {
+    if !is_valid_name(&msg.id) {
+        return Err(StdError::generic_err("Invalid atomic swap id"));
+    }
+
+    if env.message.sent_funds.is_empty() {
+        return Err(StdError::generic_err(
+            "Send some coins to create an atomic swap",
+        ));
+    }
+
+    // Ensure this is 32 bytes hex-encoded
+    let _ = parse_hex_32(&msg.hash)?;
```

Comment:
You can use the result and store it as binary (`hash: Binary::from(vec_u8)` or `hash: vec_u8.into()`)